src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt

```diff
@@ -5,26 +5,224 @@ import org.slf4j.LoggerFactory
+  }
+
+  private fun getLocationGroupByReference(reference: String): SessionLocationGroup {
+    val sessionLocationGroup = sessionLocationGroupRepository.findByReference(reference)
+      ?: throw ItemNotFoundException("SessionLocationGroup reference:$reference not found")
+    return sessionLocationGroup
+  }
+
+  private fun getSessionTemplate(reference: String): SessionTemplate {
+    val sessionTemplate = sessionTemplateRepository.findByReference(reference)
+      ?: throw TemplateNotFoundException("Template reference:$reference not found")
+    return sessionTemplate
   }
 
-  fun getSessionTemplates(sessionTemplateId: Long): SessionTemplateDto {
-    return sessionTemplateRepository.findById(sessionTemplateId).map { SessionTemplateDto(it) }
-      .orElseThrow(TemplateNotFoundException("Template id $sessionTemplateId not found"))
+  fun createSessionLocationGroup(createLocationSessionGroup: CreateLocationGroupDto): SessionLocationGroupDto {
+
+    val prison = prisonConfigService.findPrisonByCode(createLocationSessionGroup.prisonCode)
+    val sessionLocationGroup = SessionLocationGroup(
+      prison = prison,
+      prisonId = prison.id,
+      name = createLocationSessionGroup.name
+    )
+
+    val sessionLocations = createLocationSessionGroup.locations.map {
+      PermittedSessionLocation(
+        groupId = sessionLocationGroup.id,
+        sessionLocationGroup = sessionLocationGroup,
+        levelOneCode = it.levelOneCode,
+        levelTwoCode = it.levelTwoCode,
+        levelThreeCode = it.levelThreeCode,
+        levelFourCode = it.levelFourCode
+      )
+    }
+
+    sessionLocationGroup.sessionLocations.addAll(sessionLocations)
+
+    val saveSessionLocationGroup = sessionLocationGroupRepository.saveAndFlush(sessionLocationGroup)
+    return SessionLocationGroupDto(saveSessionLocationGroup)
+  }
+
+  fun updateSessionLocationGroup(reference: String, updateLocationSessionGroup: UpdateLocationGroupDto): SessionLocationGroupDto {
+    val sessionLocationGroup = getLocationGroupByReference(reference)
+    sessionLocationGroup.name = updateLocationSessionGroup.name
+    sessionLocationGroup.sessionLocations.clear()
+
+    val sessionLocations = updateLocationSessionGroup.locations.map {
+      PermittedSessionLocation(
+        groupId = sessionLocationGroup.id,
+        sessionLocationGroup = sessionLocationGroup,
+        levelOneCode = it.levelOneCode,
+        levelTwoCode = it.levelTwoCode,
+        levelThreeCode = it.levelThreeCode,
+        levelFourCode = it.levelFourCode
+      )
+    }
+
+    sessionLocationGroup.sessionLocations.addAll(sessionLocations)
+    val saveSessionLocationGroup = sessionLocationGroupRepository.saveAndFlush(sessionLocationGroup)
+    return SessionLocationGroupDto(saveSessionLocationGroup)
+  }
+
+  fun createSessionTemplate(createSessionTemplateDto: CreateSessionTemplateDto): SessionTemplateDto {
```

Comment:
not a review comment as such just a TODO. We will need to think of a few validation scenarios for create, update and delete and add tickets for each - e.g. another session at same time etc.

Reply:
makes sense